File: tp1_2025_ejer/src/ejer/StrongPasswordValidator.java
package ejer;

public class StrongPasswordValidator extends PasswordValidator {
	
    @Override
    public boolean esContraseñaValida(String password) {
        if (password == null || password.length() < 8) return false;

        if (!password.matches("[A-Za-z0-9!@#$%&*().]+")) return false;

        boolean tieneMayuscula = false;
        boolean tieneMinuscula = false;
        boolean tieneNumero = false;
        boolean tieneEspecial = false;

        for (char caracter : password.toCharArray()) {
            if (Character.isUpperCase(caracter)) tieneMayuscula = true;
            else if (Character.isLowerCase(caracter)) tieneMinuscula = true;
            else if (Character.isDigit(caracter)) tieneNumero = true;
            else if ("!@#$%&*().".indexOf(caracter) >= 0) tieneEspecial = true;

            // Si ya cumple todo, terminamos antes
            if (tieneMayuscula && tieneMinuscula && tieneNumero && tieneEspecial) return true;
        }

        return false;
    }
}

File: Python/StrongPasswordValidator.py
import re
from PasswordValidator import PasswordValidator

class StrongPasswordValidator(PasswordValidator):

    def es_contraseña_valida(self, password: str) -> bool:
        if password is None or len(password) < 8:
            return False

        if not re.fullmatch(r"[A-Za-z0-9!@#$%&*().]+", password):
            return False

        tiene_mayuscula = any(c.isupper() for c in password)
        tiene_minuscula = any(c.islower() for c in password)
        tiene_numero = any(c.isdigit() for c in password)
        tiene_especial = any(c in "!@#$%&*()." for c in password)

        return tiene_mayuscula and tiene_minuscula and tiene_numero and tiene_especial
